Mutable references are forbidden in const functions

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
const fn foo(a: &mut i32) { // { dg-error "mutable references are not allowed in const functions" }
    *a = 15;
}
```

I expected to see this happen: 
```rust
error[E0658]: mutable references are not allowed in constant functions
 --> test2.rs:1:14
  |
1 | const fn foo(a: &mut i32) {
  |              ^
  |
  = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information

error: aborting due to previous error
```

Instead, this happened: compiled without issue, even when using the value.

### Meta

- What version of Rust GCC were you using: 8ff035ddc55


## Task List

- [ ] Adding two types of context in `ConstChecker`: functions/methods and const values (and const blocks later?)
- [ ] Check when a type is being visited: If we're on a mutable reference and the context is a function/method, error out

Testcase available on https://github.com/CohenArthur/gccrs/tree/forbid-mut-refs-in-const-fns

--- a/gcc/rust/checks/errors/rust-const-checker.cc
+++ b/gcc/rust/checks/errors/rust-const-checker.cc
@@ -898,8 +898,12 @@ ConstChecker::visit (RawPointerType &)
 {}
 
 void
-ConstChecker::visit (ReferenceType &)
-{}
+ConstChecker::visit (ReferenceType &type)
+{
+  if (const_context.is_in_context () && type.is_mut ())
+    rust_error_at (type.get_locus (),
+		   "mutable references are not allowed in constant functions");
+}
 
 void
 ConstChecker::visit (ArrayType &type)

--- a/gcc/testsuite/rust/compile/const10.rs
+++ b/gcc/testsuite/rust/compile/const10.rs
@@ -1,0 +1,3 @@
+const fn foo (a: &mut i32) { // { dg-error "mutable references are not allowed in constant functions" }
+	*a = 1;
+}

